monai/apps/auto3dseg/ensemble_builder.py

```diff
@@ -0,0 +1,227 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import json
+import logging
+import os
+from abc import ABC, abstractmethod
+from copy import deepcopy
+from typing import Any, Dict, List, Optional, Sequence
+
+import numpy as np
+
+from monai.apps.auto3dseg.bundle_gen import BundleAlgo
+from monai.auto3dseg import concat_val_to_np
+from monai.bundle import ConfigParser
+from monai.utils.enums import StrEnum
+
+logger = logging.getLogger(__name__)
+
+
+class ScriptEnsembleKeys(StrEnum):
+    """
+    Default keys for Mixed Ensemble
+    """
+
+    ID = "identifier"
+    ALGO = "infer_algo"
+    SCORE = "best_metric"
+
+
+class ScriptEnsemble(ABC):
+    """
+    The base class of Ensemble methods
+    """
+
+    def __init__(self):
+        self.algos = []
+        self.infer_files = []
+
+    def set_algos(self, infer_algos):
+        """
+        Register model in the ensemble
+        """
+        self.algos = deepcopy(infer_algos)
+
+    def get_algo(self, identifier):
+        """
+        Get a model by identifier.
+
+        Args:
+            identifier: the name of the bundleAlgo
+        """
+        for algo in self.algos:
+            if identifier == algo[ScriptEnsembleKeys.ID]:
+                return algo
+
+    def set_infer_files(self, dataroot: str, data_src_cfg_file: str):
+        """
+        Set the files to perform model inference.
+
+        Args:
+            dataroot: the path of the files
+            data_src_cfg_file: the data source file path
+        """
+        with open(data_src_cfg_file) as f:
+            datalist = json.load(f)
+
+        for d in datalist["testing"]:
+            self.infer_files.append({"image": os.path.join(dataroot, d["image"])})
+
+    @abstractmethod
+    def rank_algos(self):
+        raise NotImplementedError
+
+    @abstractmethod
+    def predict(self):
+        """
+        predict results after the models are ranked/weighted
+        """
+        raise NotImplementedError
+
+
+class EnsembleBestN(ScriptEnsemble):
+    """
+    Ensemble method that select N model out of all using the models' best_metric scores
+
+    Args:
+        n_best: number of models to pick for ensemble (N).
+    """
+
+    def __init__(self, n_best: int = 5):
+
+        super().__init__()
+        self.n_best = n_best
+
+    def sort_score(self):
+        """sort the best_metrics"""
+        scores = concat_val_to_np(self.algos, [ScriptEnsembleKeys.SCORE])
+        return np.argsort(scores).tolist()
+
+    def rank_algos(self):
```

Comment:
```suggestion
    def rank_algos(self, n_best: = -1):
```

Reply:
Incorporated the change by manual replacing. thanks